catalyst-gateway-crates/c509-certificate/src/c509_algo_identifier.rs

```diff
@@ -0,0 +1,87 @@
+//! C509 Algorithm Identifier
+//! This module handle the `AlgorithmIdentifier` type where OID does not fall into the
+//! table.
+//!
+//! ```cddl
+//!    AlgorithmIdentifier = int / ~oid / [ algorithm: ~oid, parameters: bytes ]
+//! ```
+//!
+//! For more information about `AlgorithmIdentifier`,
+//! visit [C509 Certificate](https://datatracker.ietf.org/doc/draft-ietf-cose-cbor-encoded-cert/09/)
+
+use asn1_rs::Oid;
+use minicbor::{encode::Write, Decode, Decoder, Encode, Encoder};
+
+use crate::c509_oid::C509oid;
+
+/// A struct represents the `AlgorithmIdentifier` type.
+#[derive(Debug, Clone, PartialEq)]
+pub struct AlgorithmIdentifier {
+    /// A `C509oid`
+    oid: C509oid,
+    /// An optional parameter string
+    param: Option<String>,
+}
+
+impl AlgorithmIdentifier {
+    /// Create new instance of `AlgorithmIdentifier`.
+    #[must_use]
+    pub fn new(oid: Oid<'static>, param: Option<String>) -> Self {
+        Self {
+            oid: C509oid::new(oid),
+            param,
+        }
+    }
+
+    /// Get the OID.
+    pub(crate) fn get_oid(self) -> Oid<'static> {
+        self.oid.get_oid()
+    }
+
+    /// Get the parameter.
+    pub(crate) fn get_param(self) -> Option<String> {
+        self.param
+    }
```

Comment:
Shouldn't comsume itself just to read a value. You can make it returns a reference of a value with `&` instead.